lib/Synchronise/Network/Server.hs

```diff
@@ -212,9 +230,137 @@ notify (RequestChange note) = do
             throwError UnknownKeyError
         Just _ ->
             liftIO . addWork store . WorkNotify $ ForeignKey ent_name src_name fid
+
     return ResponseChange
 
+
 --------------------------------------------------------------------------------
+
+-- * Asynchronous Server Workers
+
+-- | Get a work item from the work queue, apply a constant backoff if there is
+--   nothing in the queue.
+--
+--   This function must not
+getWorkBackoff :: Store store => store -> IO (WorkItemID, WorkItem)
+getWorkBackoff store = do
+  work <- getWork store
+  case work of
+    Nothing -> threadDelay 50000 >> getWorkBackoff store
+    Just x  -> return x
+
+-- | A worker for the synchronised server.
+--   These workers cannot die, they simply log any errors and keep going.
+--
+worker :: Store store => store -> Configuration -> IO ()
+worker store cfg = go
+  where go = do
+          (_, item) <- liftIO $ getWorkBackoff store
+          case item of
+            WorkNotify fk -> do
+              liftIO . debugM logName $ "Processing a notifcation: " <> show fk
+              processNotification store cfg fk
+            WorkApplyPatch did diff -> do
+              liftIO . debugM logName $ "Processing a diff: " <> show did
+              processDiff did diff
+          go
```

Comment:
This must never happen. In retcon there's a state machine in a column which is updated in a transaction to "claim" a piece of work. The claim either expires (because the worker crashed) or it eventually finishes the work item and deletes the record.

Also in retcon, the "do some work" operation is something like this:

`processWork :: store -> (WorkItem -> IO (Maybe Error)) -> IO ()`

So the store does the transactions completely inside `processWork`. I'm happy with getWork/completeWork though; just need to get the state machine in there so we can do it safely.